### ISSUE TITLE
poly-elixir-web-mode config works, but intermittent or only after some trigger etc

### ISSUE THREAD

**Issue by anildigital** [polymode/polymode#308]
Hello, I use following config (poly-elixir-web-mode) to use for Surface templates in my Elixir files 

```elisp
;; Assumes web-mode and elixir-mode are already set up
;;
(use-package
  polymode
  :ensure t
  :mode ("\.ex$" . poly-elixir-web-mode)
  :hook (elixir-mode . poly-elixir-web-mode)
  :config (define-hostmode poly-elixir-hostmode
            :mode 'elixir-mode)
  (define-innermode poly-surface-expr-elixir-innermode
    :mode 'web-mode
    :head-matcher (rx line-start (* space) "~F" (= 3 (char "\"'")) line-end)
    :tail-matcher (rx line-start (* space)
                      (= 3 (char "\"'")) line-end)
    :head-mode 'host
    :tail-mode 'host
    :allow-nested nil
    :keep-in-mode 'host
    :fallback-mode 'host)
  (define-polymode
    poly-elixir-web-mode
    :hostmode 'poly-elixir-hostmode
    :innermodes '(poly-surface-expr-elixir-innermode)))
```

Here is example Elixir code where this need to work for F tag

```elixir
defmodule MyAppFlaskWeb.Components.GameStats do
  use MyAppFlaskWeb, :stateful_component

  alias Foo.Autolayouts.{LeftToRight, TopToDown}
  alias Foo.Components.{Heading, Text}
  alias Foo.Icons.{ArrowsBottomRight, ArrowsTopRight}

  def render(assigns) do
    ~F"""
    <TopToDown class="game_stats" gap="gap-4">
      <Heading size={20} is_regular class="sm:hidden">{gettext("stats")}</Heading>
      <Heading size={24} is_regular class="hidden sm:block">{gettext("stats")}</Heading>

      <div class="grid grid-cols-2 sm:grid-cols-4 gap-4">
        <LeftToRight class="justify-between pb-4 border-b" centered>
          <Text size={14} color="trunks-100">ABC</Text>
          <LeftToRight gap="gap-2" centered>
            <Text size={14}>11.08%</Text>
          </LeftToRight>
        </LeftToRight>
      </div>
    </TopToDown>
    """
  end
end
```

I notice, when I open files, it doesn't show syntax highlighting, I had to hit enter and delete back after `~F"""` or I trigger `swiper` and rendering works. But it's very intermittent. I wonder why is that.

Thanks

**Comment by anildigital** [comment]
I also see this when I try to reopen the file

`(jit-lock--run-functions 280 497) [span 280 497  session_token_enforcer.ex[web]] -> (font-lock-default-fontify-region 280 497): Args out of range: 1, 1`

**Comment by anildigital** [comment]
Update I am using following config and still facing exact same issues

```elisp
(use-package
   polymode
  :ensure t
  :mode ("\\.ex\\'" . poly-elixir-web-mode)
  :init (setq web-mode-engines-alist '(("elixir" . "\\.ex\\'")))
  :config
  (define-hostmode poly-elixir-hostmode :mode 'elixir-mode)
  (define-innermode poly-surface-expr-elixir-innermode
    :mode 'web-mode
    :head-matcher (rx line-start (* space) "~F" (= 3 (char "\"'")) line-end)
    :tail-matcher (rx line-start (* space) (= 3 (char "\"'")) line-end)
    :head-mode 'host
    :tail-mode 'host
    :allow-nested nil
    :keep-in-mode 'host
    :fallback-mode 'host)
  (define-polymode poly-elixir-web-mode
    :hostmode 'poly-elixir-hostmode
    :innermodes '(poly-surface-expr-elixir-innermode)))
```

**Comment by vspinu** [comment]
Thanks for the detailed report. Is this a recent issue for you?

**Comment by anildigital** [comment]
@vspinu Not a recent one, it's been like few months, this issue exists.

**Comment by vspinu** [comment]
Merged upstream. Could you please try and see if it's working for you?

**Comment by anildigital** [comment]
@vspinu Works perfectly now. Thanks a lot! ❤️❤️❤️

**Comment by d3caf** [comment]
Just posting here for visibility, I'm using @anildigital's config on HEAD and updated web-mode and am still seeing the issue described here. No jit fontify errors though.

**Comment by vspinu** [comment]
Intermitent and no errors whatsoever in the `*Messages*` buffer? Which emacs is this?

**Comment by d3caf** [comment]
This is from a clean start of Doom. The buffer I have open actually shows no web-mode highlighting within my `~F` block.

Running emacs-plus@28 natively compiled.

![CleanShot 2022-01-19 at 14 45 28](https://user-images.githubusercontent.com/18057119/150202877-2dcbd61c-5865-4110-bfc1-ea8e51c914a2.png)

**Comment by vspinu** [comment]
It's probably interaction with something else. Maybe evil ? :thinking:I have just tested with emacses all the way back to 25.2. Works as expected. I am closing this as for now. If you could isolate it down to something reproducible please open a new issue.

**Comment by d3caf** [comment]
@vspinu just an update on this, I can get highlighting from polymode to work correctly if I fiddle around with enabling and disabling my elixir-web polymode and font-lock. I haven't quite figured out the correct order to do things, but it does seem to be font-lock related.

**Comment by vspinu** [comment]
I am not quite happy with how font-lock is installed on startup. I have in plan to revisit how it is done. If you could replicate your issues with some minimal setup that would be of considerable help. For instance from the root of this repo you can do `make elpa` then `make start` to start the minimal env with polymode. The file `targets/deps` contains a list of extra libraries that you would like to fetch from elpa.
